pymc3/step_methods/particle.py

```diff
@@ -0,0 +1,73 @@
+import theano
+from .arraystep import ArrayStepShared
+from ..model import modelcontext
+from ..theanof import inputvars, make_shared_replacements
+from ..blocking import ArrayOrdering, DictToArrayBijection
+from emcee import EnsembleSampler
+import pymc3 as pm
+import numpy as np
+
+class ParticleStep(ArrayStepShared):
+    def __new__(cls, *args, **kwargs):
+        obj = ArrayStepShared.__new__(cls, *args[1:], **kwargs)
+        return obj
+
+    def __init__(self, nparticles=None, model=None, mode=None, **kwargs):
+        model = modelcontext(model)
+        vars = model.vars
+        vars = inputvars(vars)
+        shared = make_shared_replacements(vars, model)
+        self.mode = mode
+        self.model = model
+        self.vars = vars
+        self.dim = sum([i.dsize for i in self.vars])
+        if nparticles is None:
+            self.nparticles = (self.dim*2) + 2
+        else:
+            self.nparticles = nparticles
+        self.ordering = ArrayOrdering(vars, self.nparticles)
+        self.shared = {str(var): shared for var, shared in shared.items()}
+        self.blocked = True
+        self.t_func = logp(model.logpt, vars, shared)
+
+
+    def step(self, point):
+        for var, share in self.shared.items():
+            share.container.storage[0] = point[var]
+
+        bij = DictToArrayBijection(self.ordering, point)
+        apoint = self.astep(bij.map(point))
+        return bij.rmap(apoint)
+
+
+class EmceeSamplerStep(ParticleStep):
```

Comment:
May want to avoid `Emcee` in the name, to avoid giving the impression that we are using emcee under the hood. Perhaps `AffineInvariantEnsemble` or `GoodmanWeare`?